src/h5web/visualizations/shared/utils.ts

```diff
@@ -1,6 +1,35 @@
 import { scaleLinear } from 'd3-scale';
+import { CSSProperties } from 'react';
+import { AxisOffsets } from './models';
 
 export const adaptedNumTicks = scaleLinear()
   .domain([300, 900])
   .rangeRound([3, 10])
   .clamp(true);
+
+export function computeVisSize(
+  visAreaSize: { width: number; height: number },
+  axisOffsets: AxisOffsets,
+  aspectRatio?: number
+): CSSProperties | undefined {
+  const { width, height } = visAreaSize;
+
+  if (width === 0 && height === 0) {
+    return undefined;
+  }
+
+  if (!aspectRatio) {
+    return { width, height };
+  }
+
+  const availableWidth = width - axisOffsets.left;
+  const availableHeight = height - axisOffsets.bottom;
+
+  // Determine how to compute canvas size to fit available space while maintaining aspect ratio
+  const shouldAdjustWidth = availableWidth >= availableHeight * aspectRatio;
+  const canvasSize = shouldAdjustWidth
+    ? { width: availableHeight * aspectRatio, height: availableHeight }
+    : { width: availableWidth, height: availableWidth * aspectRatio };
+
+  return { ...canvasSize, boxSizing: 'content-box' };
```

Comment:
Just realised that now that the padding lives in the parent, you could move `content-box` there as well. It can be applied regardless of the aspect ratio config.